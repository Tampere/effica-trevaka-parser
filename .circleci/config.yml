# SPDX-FileCopyrightText: 2021 City of Tampere
#
# SPDX-License-Identifier: LGPL-2.1-or-later

version: 2.1

orbs:
  node: circleci/node@5.2.0

executors:
  reuse_tool:
    docker:
      - image: fsfe/reuse:0.13.0 # NOTE: Update bin/add-license-headers.sh to match
    resource_class: small
  machine:
    machine:
      image: ubuntu-2204:2024.01.1
    resource_class: medium
    environment:
      BUILDKIT_PROGRESS: plain

commands:
  checkout_with_submodules:
    steps:
      - checkout
      - run: git submodule sync
      - run: git submodule update --init --recursive

jobs:
  check_licenses:
    executor: reuse_tool
    steps:
      - checkout
      - run:
          name: reuse lint
          command: |
            reuse lint || {
              echo '';
              echo '=========================================';
              echo 'Please run: ./bin/add-license-headers.sh';
              echo '=========================================';
              exit 1
            }
  test:
    executor: machine
    steps:
      - checkout_with_submodules
      - run:
          name: Start up compose stack
          command: docker-compose -f docker-compose-it-db.yml up -d --wait --build
      - run:
          name: Waiting for compose stack to be up...
          command: docker-compose -f docker-compose-it-db.yml wait flyway
      - node/install-packages:
          include-branch-in-cache-key: false
      - run: npm run integration-test
      - run:
          name: Collect docker-compose logs
          command: docker-compose -f docker-compose-it-db.yml logs --tail=all > /tmp/docker-compose-logs.txt
          when: always
      - store_artifacts:
          path: /tmp/docker-compose-logs.txt
          destination: docker-compose-logs.txt

workflows:
  version: 2
  test:
    jobs:
      - check_licenses
      - test
