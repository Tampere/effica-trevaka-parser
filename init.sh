#!/bin/bash

# SPDX-FileCopyrightText: 2023-2024 Tampere region
#
# SPDX-License-Identifier: LGPL-2.1-or-later

set -e
set -u

function create_database() {
  psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" <<-EOSQL
    CREATE DATABASE $1;
    \connect $1
    CREATE SCHEMA migration;
EOSQL
}

create_database vesilahti
