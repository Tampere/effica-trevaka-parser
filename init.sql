-- SPDX-FileCopyrightText: 2021 City of Tampere
--
-- SPDX-License-Identifier: LGPL-2.1-or-later

-- Add extensions here as needed (for UUIDs at least)
CREATE SCHEMA IF NOT EXISTS ext;
CREATE SCHEMA IF NOT EXISTS migration;

CREATE EXTENSION IF NOT EXISTS "uuid-ossp" WITH SCHEMA ext;
CREATE EXTENSION IF NOT EXISTS btree_gist WITH SCHEMA ext;