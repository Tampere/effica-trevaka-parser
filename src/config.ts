export const config = {
    cityVariant: process.env.CITY_VARIANT ?? "tampere",
    port: process.env.PORT ?? 3000,
    isTimed: process.env.ISTIMED?.toUpperCase() === "TRUE",
    migrationSchema: process.env.MIGRATION_SCHEMA ?? "migration",
    extensionSchema: process.env.EXTENSION_SCHEMA ?? "ext",
    migrationDb: {
        host: process.env.PGHOST ?? "localhost",
        port: 5432,
        user: process.env.PGUSER ?? "postgres",
        password: process.env.PGPASSWORD ?? "postgres",
        database: process.env.PGDATABASE ?? "migration",
    },
    xmlParserOptions: {
        parseNodeValue: true,
        arrayMode: true,
        trimValues: true,
        parseTrueNumberOnly: true
    },
    csvParserOptions: {
        trim: true,
        delimiter: "|"
    }
}