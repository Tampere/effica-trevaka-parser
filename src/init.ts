// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import migrationDb from "./db/db";
import { baseQueryParameters, runQueryFile } from "./util/queryTools";

export const initDb = async () => {
    return await migrationDb.tx(async (t) => {
        await runQueryFile("init.sql", t, baseQueryParameters);
        await runQueryFile("functions.sql", t, baseQueryParameters);
    });
};
