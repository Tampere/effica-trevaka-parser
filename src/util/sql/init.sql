-- SPDX-FileCopyrightText: 2021 City of Tampere
--
-- SPDX-License-Identifier: LGPL-2.1-or-later

CREATE TABLE IF NOT EXISTS ${migrationSchema:name}.idmap (
    type TEXT NOT NULL,
    effica_guid TEXT NOT NULL,
    evaka_id UUID NOT NULL,
    created TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
    updated TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT now(),
    PRIMARY KEY (type, effica_guid)
);
