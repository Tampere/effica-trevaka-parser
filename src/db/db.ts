import pgPromise from "pg-promise"
import { config } from "../config"

export const pgp = pgPromise({})

// date parsing seems to be scuffed, does local tz transformation but declares it UTC -> corrupts dates
//pgp.pg.types.setTypeParser(1082, dateParser)

const migrationDb = pgp({ ...config.migrationDb })

export default migrationDb