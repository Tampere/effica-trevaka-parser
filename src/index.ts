// SPDX-FileCopyrightText: 2021 City of Tampere
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import app from "./app"
import { config } from "./config"

app.listen(config.port, () => {
    console.log(`Parser app listening at http://localhost:${config.port}`)
})

