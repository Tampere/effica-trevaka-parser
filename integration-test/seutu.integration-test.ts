// SPDX-FileCopyrightText: 2023-2024 Tampere region
//
// SPDX-License-Identifier: LGPL-2.1-or-later

import { initDb } from "../src/init";
import { config as configFromEnv } from "../src/config";
import { pgp } from "../src/db/db";
import { cleanupDb } from "../src/util/testTools";
import {
    AREA_COLUMN_SET, findTransferFridge,
    findTransferIncomes,
    imports,
    transfers,
    transforms,
    UNIT_COLUMN_SET
} from "./test-utils";

const config = {
    ...configFromEnv,
    cityVariant: "seutu",
    migrationDb: { ...configFromEnv.migrationDb, database: "evaka_local" },
};
const db = pgp(config.migrationDb);

beforeAll(() => db.tx(initDb));
beforeEach(() => db.tx((tx) => cleanupDb(tx, true)));
afterAll(db.$pool.end);
const importBasePath = `${__dirname}/data/sem/seutu`;

describe("seutu", () => {
    beforeEach(async () => {
        await db.none(pgp.helpers.insert(areas, AREA_COLUMN_SET));
        await db.none(pgp.helpers.insert(units, UNIT_COLUMN_SET));
    });

    it("pre", async () => {
        await imports(
            db,
            {
                path: `${importBasePath}/pre-test/person`,
                importTarget: "effica_person",
            },
            {
                path: `${importBasePath}/pre-test/fridge/child`,
                importTarget: "effica_fridge_child",
            },
            {
                path: `${importBasePath}/pre-test/placement`,
                importTarget: "effica_placement",
            },
            {
                path: `${importBasePath}/pre-test/income`,
                importTarget: "effica_income",
            },
        );
        await transforms(db, config, "pre-1", "pre-2");
        await transfers(db, "pre-1", "pre-2");
    });

    it("fridge", async () => {
        await imports(
            db,
            {
                path: `${importBasePath}/fridge-test/person`,
                importTarget: "effica_person",
            },
            {
                path: `${importBasePath}/fridge-test/fridge/child`,
                importTarget: "effica_fridge_child",
            },
            {
                path: `${importBasePath}/fridge-test/income`,
                importTarget: "effica_income",
            },
        );
        await transforms(db, config, "person", "fridge");
        await transfers(db, "person", "fridge");
        expect(await db.tx(findTransferFridge)).toMatchSnapshot();
    });

    it("income", async () => {
        await imports(
            db,
            {
                path: `${importBasePath}/income-test/person`,
                importTarget: "effica_person",
            },
            {
                path: `${importBasePath}/income-test/income`,
                importTarget: "effica_income",
            },
        );
        await transforms(db, config, "person", "income");
        await transfers(db, "person", "income");
        expect(await db.tx(findTransferIncomes)).toMatchSnapshot();
    });

    it("post", async () => {
        await imports(
            db,
            {
                path: `${importBasePath}/post-test/person`,
                importTarget: "effica_person",
            },
            {
                path: `${importBasePath}/post-test/fridge/child`,
                importTarget: "effica_fridge_child",
            },
            {
                path: `${importBasePath}/post-test/income`,
                importTarget: "effica_income",
            },
        );
        await transforms(db, config, "person", "fridge");
        await transforms(db, config, "post");
        await transfers(db, "post");
    });
});

const areas = [
    {
        id: "d297bf95-dea7-4e9a-8ecb-5d6c9a85b850",
        name: "Alue A",
    },
];

const units = [
    {
        id: "fcba1717-ea3d-4611-8ea5-32669d630ee9",
        name: "P채iv채koti A",
        care_area_id: "d297bf95-dea7-4e9a-8ecb-5d6c9a85b850",
        provider_type: "MUNICIPAL",
    },
    {
        id: "794ae905-982d-42aa-93d1-2dfa6b9938b7",
        name: "P채iv채koti B",
        care_area_id: "d297bf95-dea7-4e9a-8ecb-5d6c9a85b850",
        provider_type: "MUNICIPAL",
    },
];
