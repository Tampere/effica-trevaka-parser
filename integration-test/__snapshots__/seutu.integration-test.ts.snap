// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seutu fridge 1`] = `
[
  {
    "child_social_security_number": "010124A920A",
    "end_date": "2041-12-31",
    "head_of_child_social_security_number": "010106A914F",
    "start_date": "2024-01-01",
  },
]
`;

exports[`seutu income 1`] = `
[
  {
    "data": {
      "MAIN_INCOME": {
        "amount": 524165,
        "coefficient": "MONTHLY_NO_HOLIDAY_BONUS",
        "monthlyAmount": 524165,
        "multiplier": 1,
      },
    },
    "effect": "INCOME",
    "evaka_ssn": "010106A914F",
    "is_entrepreneur": false,
    "notes": "Tulojen summa Efficasta",
    "valid_from": "2024-01-01",
    "valid_to": null,
    "works_at_echa": false,
  },
  {
    "data": {
      "MAIN_INCOME": {
        "amount": 402000,
        "coefficient": "MONTHLY_NO_HOLIDAY_BONUS",
        "monthlyAmount": 402000,
        "multiplier": 1,
      },
    },
    "effect": "INCOME",
    "evaka_ssn": "010124A920A",
    "is_entrepreneur": false,
    "notes": "Tulojen summa Efficasta",
    "valid_from": "2024-01-01",
    "valid_to": null,
    "works_at_echa": false,
  },
]
`;
